src/main/java/hudson/plugins/s3/S3BucketPublisher.java

```diff
@@ -436,6 +439,49 @@ private String getFilename(FilePath src, boolean flatten, int searchIndex) {
         return fileName;
     }
 
+    @Extension
+    public static final class S3DeletedItemListener extends ItemListener {
+        @Override
+        public void onDeleted(Item item) {
+            if (item instanceof Job<?, ?> job) {
+                handleJobDeletion(job);
+            } else if (item instanceof ItemGroup<?> itemGroup) {
+                handleItemGroupDeletion(itemGroup);
+            }
+        }
+
+        private void handleJobDeletion(Job<?, ?> job) {
+            for (Run<?, ?> run : job.getBuilds()) {
+                S3ArtifactsAction artifacts = run.getAction(S3ArtifactsAction.class);
+                if (artifacts != null) {
+                    S3Profile profile = S3BucketPublisher.getProfile(artifacts.getProfile());
+                    if (profile.isDeleteArtifactsRecursively()) {
+                        for (FingerprintRecord record : artifacts.getArtifacts()) {
+                            if (!record.isKeepForever()) {
+                                try {
+                                    profile.delete(run, record);
+                                } catch (Exception e) {
+                                    Logger.getLogger(S3DeletedItemListener.class.getName())
+                                            .log(Level.WARNING, "Failed to delete S3 artifact: " + record.getName(), e);
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        private void handleItemGroupDeletion(ItemGroup<?> itemGroup) {
+            for (Item item : itemGroup.getItems()) {
+                if (item instanceof Job<?, ?>) {
+                    handleJobDeletion((Job<?, ?>) item);
+                } else if (item instanceof ItemGroup<?>) {
+                    handleItemGroupDeletion((ItemGroup<?>) item);
+                }
+            }
```

Comment:
You can also use the pattern matching here (which is great btw). I wasn't sure it was "authorized" (which version of Java is supported by core 2.479.3) and which version the pattern matching was introduced. But all good ([1](https://www.jenkins.io/doc/book/platform-information/support-policy-java/#running-jenkins-system) and [2](https://openjdk.org/jeps/394))

```suggestion
            for (Item item : itemGroup.getItems()) {
                if (item instanceof Job<?, ?> job) {
                    handleJobDeletion(job);
                } else if (item instanceof ItemGroup<?> itemGroup) {
                    handleItemGroupDeletion(itemGroup);
                }
            }
```

Reply:
@alecharp thanks for pointing out

I've applied your change, the only thing is that because `itemGroup` was already used as an argument for the method `handleItemGroupDeletion()` I've used `childItemGroup` as a pattern variable.